Panzer: MiniEM example compilation error in Cuda build - calling host function from device function

## Bug Report
@trilinos/panzer 

### Description
While building Panzer with examples turned on in a Cuda build with cuda/11.2.2 and the Kokkos develop branch, compilation errors occurred in the mini-em example:

Here are warnings and error following the compile line from `make -j1 VERBOSE=1` 
```
/ascldap/users/ndellin/trilinos/Trilinos/packages/panzer/mini-em/src/solvers/MiniEM_FullMaxwellPreconditionerFactory.hpp(18): warning: function "Teko::PreconditionerFactory::buildPreconditionerOperator(Teko::LinearOp &, Teko::PreconditionerState &) const" is hidden by "mini_em::FullMaxwellPreconditionerFactory::buildPreconditionerOperator" -- virtual function override intended?

/ascldap/users/ndellin/trilinos/Trilinos/packages/panzer/mini-em/src/solvers/MiniEM_HigherOrderMaxwellPreconditionerFactory.hpp(18): warning: function "Teko::PreconditionerFactory::buildPreconditionerOperator(Teko::LinearOp &, Teko::PreconditionerState &) const" is hidden by "mini_em::HigherOrderMaxwellPreconditionerFactory::buildPreconditionerOperator" -- virtual function override intended?

/ascldap/users/ndellin/trilinos/Trilinos/packages/panzer/mini-em/src/solvers/MiniEM_FullMaxwellPreconditionerFactory_Augmentation.hpp(18): warning: function "Teko::PreconditionerFactory::buildPreconditionerOperator(Teko::LinearOp &, Teko::PreconditionerState &) const" is hidden by "mini_em::FullMaxwellPreconditionerFactory_Augmentation::buildPreconditionerOperator" -- virtual function override intended?

/ascldap/users/ndellin/trilinos/Trilinos/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp(81): error: calling a __host__ function("shards::CellTopology::~CellTopology") from a __device__ function("Kokkos::Impl::ParallelFor< ::Kokkos::Impl::ViewValueFunctor< ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::shards::CellTopology, (bool)0> ,  ::Kokkos::RangePolicy< ::Kokkos::Cuda,  ::Kokkos::IndexType<long> ,     ::Kokkos::Impl::ViewValueFunctor< ::Kokkos::Device< ::Kokkos::Cuda,  ::Kokkos::CudaUVMSpace> ,  ::shards::CellTopology, (bool)0> ::DestroyTag > ,  ::Kokkos::Cuda> ::operator () const") is not allowed

/ascldap/users/ndellin/trilinos/Trilinos/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp(81): error: identifier "shards::CellTopology::~CellTopology" is undefined in device code

2 errors detected in the compilation of "/ascldap/users/ndellin/trilinos/Trilinos/packages/panzer/mini-em/example/BlockPrec/main.cpp".
make[2]: *** [packages/panzer/mini-em/example/BlockPrec/CMakeFiles/PanzerMiniEM_BlockPrec.dir/build.make:75: packages/panzer/mini-em/example/BlockPrec/CMakeFiles/PanzerMiniEM_BlockPrec.dir/main.cpp.o] Error 1
```

Possibly a missing `KOKKOS_INLINE_FUNCTION` marking is missing on a function to give the proper `__host__ __device__` annotation?

### Steps to Reproduce
SHA1:  Built with changes from PR #11370 

Aside from enabling Cuda and Panzer's examples in the build, I'm not sure if these configuration options may be relevant

```

      -D Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=ON \
      -D Kokkos_ARCH_VOLTA70=ON \
      -D Kokkos_ARCH_POWER9=ON \
      -D Kokkos_ENABLE_CUDA=ON \
      -D Kokkos_ENABLE_CUDA_LAMBDA=ON \
      -D Kokkos_ENABLE_CUDA_UVM=ON \
      -D TPL_ENABLE_CUSPARSE:BOOL=OFF \
      -D Sacado_ENABLE_HIERARCHICAL_DFAD:BOOL=ON \
      -D Teuchos_ENABLE_DEBUG_RCP_NODE_TRACING=ON \
      -D Trilinos_AUTOGENERATE_TEST_RESOURCE_FILE=ON \

```

--- a/packages/panzer/mini-em/src/solvers/MiniEM_DiscreteCurl.hpp
+++ b/packages/panzer/mini-em/src/solvers/MiniEM_DiscreteCurl.hpp
@@ -165,15 +165,14 @@ void addDiscreteCurlToRequestHandler(
     Kokkos::deep_copy(refDofCoeffs_h, refDofCoeffs);
     // set up the topology of each face in an element for computing DOF coefficients
     // in 2D coefficients are same as reference coefficients
-    typename Kokkos::DynRankView<shards::CellTopology,DeviceSpace>::HostMirror
-      sub_topologies(Kokkos::ViewAllocateWithoutInitializing("sub_topologies"), hdivCardinality);
+    shards::CellTopology sub_topologies[hdivCardinality];
     if(dim < 3)
       for(int i = 0; i < hdivCardinality; i++)
         dofCoeffs_h(0,i) = refDofCoeffs_h(i);
     else {
       for(int iface = 0; iface < hdivCardinality; iface++){
         shards::CellTopology sub_topology(topology.getCellTopologyData(dim-1,iface));
-        sub_topologies(iface) = sub_topology;
+        sub_topologies[iface] = sub_topology;
       }
     }
 
@@ -217,7 +216,7 @@ void addDiscreteCurlToRequestHandler(
         if(dim==3){
           elemOrts(0).getFaceOrientation(fOrt.data(),hdivCardinality);
           for(int iface = 0; iface < hdivCardinality; iface++){
-            Intrepid2::Impl::OrientationTools::getJacobianOfOrientationMap(ortJacobian, sub_topologies(iface), fOrt(iface));
+            Intrepid2::Impl::OrientationTools::getJacobianOfOrientationMap(ortJacobian, sub_topologies[iface], fOrt(iface));
             auto ortJacobianDet = ortJacobian(0,0)*ortJacobian(1,1)-ortJacobian(1,0)*ortJacobian(0,1);
             for(int idim = 0; idim < dim; idim++)
               dofCoeffs_h(0,iface,idim) = refDofCoeffs_h(iface,idim)*ortJacobianDet;
@@ -384,15 +383,14 @@ void addDiscreteCurlToRequestHandler(
     Kokkos::deep_copy(refDofCoeffs_h, refDofCoeffs);
     // set up the topology of each face in an element for computing DOF coefficients
     // in 2D coefficients are same as reference coefficients
-    typename Kokkos::DynRankView<shards::CellTopology,DeviceSpace>::HostMirror
-      sub_topologies(Kokkos::ViewAllocateWithoutInitializing("sub_topologies"), hdivCardinality);
+    shards::CellTopology sub_topologies[hdivCardinality];
     if(dim < 3)
       for(int i = 0; i < hdivCardinality; i++)
         dofCoeffs_h(0,i) = refDofCoeffs_h(i);
     else {
       for(int iface = 0; iface < hdivCardinality; iface++){
         shards::CellTopology sub_topology(topology.getCellTopologyData(dim-1,iface));
-        sub_topologies(iface) = sub_topology;
+        sub_topologies[iface] = sub_topology;
       }
     }
 
@@ -424,7 +422,7 @@ void addDiscreteCurlToRequestHandler(
         if(dim==3){
           elemOrts(0).getFaceOrientation(fOrt.data(),hdivCardinality);
           for(int iface = 0; iface < hdivCardinality; iface++){
-            Intrepid2::Impl::OrientationTools::getJacobianOfOrientationMap(ortJacobian, sub_topologies(iface), fOrt(iface));
+            Intrepid2::Impl::OrientationTools::getJacobianOfOrientationMap(ortJacobian, sub_topologies[iface], fOrt(iface));
             auto ortJacobianDet = ortJacobian(0,0)*ortJacobian(1,1)-ortJacobian(1,0)*ortJacobian(0,1);
             for(int idim = 0; idim < dim; idim++)
               dofCoeffs_h(0,iface,idim) = refDofCoeffs_h(iface,idim)*ortJacobianDet;

